File: solutions/c/beer-song/1/beer_song.c
// for Shree DR.MDD
#include "beer_song.h"
#include "beer_song.h"
#include <stdio.h>

static const char * L1[] = {
    "No more bottles of beer on the wall, no more bottles of beer.",
    "1 bottle of beer on the wall, 1 bottle of beer.",
    "%u bottles of beer on the wall, %u bottles of beer."
};

static const char * L2[] = {
    "Go to the store and buy some more, 99 bottles of beer on the wall.",
    "Take it down and pass it around, no more bottles of beer on the wall.",
    "Take one down and pass it around, 1 bottle of beer on the wall.",
    "Take one down and pass it around, %u bottles of beer on the wall."
};

void recite(uint8_t start, uint8_t drop, char **msg) {
    uint8_t last = start - (drop - 1);
    for (int i = start; i >= last; i--) {
        sprintf(*msg++, L1[i > 1 ? 2 : i], i, i);
        if (i == 0) {
            sprintf(*msg++, L2[0]);
        } else if (i == 1) {
            sprintf(*msg++, L2[1]);
        } else if (i == 2) {
            sprintf(*msg++, L2[2]);
        } else {
            sprintf(*msg++, L2[3], i - 1);
        }
        msg++;
    }
}

File: solutions/delphi/beer-song/1/beer_song.pas
unit uBeerSong;

interface

type
  Beer = class
  private
    class function NumBottles(const AIndex:integer):string;
    class function IsLastBottle(const AIndex:integer):string;
  public
    class function Recite(const startBottles: integer; const takeDown: integer):string;
  end;

implementation

uses
  System.SysUtils;

class function Beer.NumBottles(const AIndex:integer):string;
begin
  if AIndex > 1 then
    Result := AIndex.ToString + ' bottles'
  else if AIndex = 1 then
    Result := AIndex.ToString + ' bottle'
  else if
    AIndex = 0 then Result := 'no more bottles';
end;

class function Beer.IsLastBottle(const AIndex:integer):string;
begin
  if AIndex = 1 then Result := 'it'
  else Result := 'one';
end;

class function Beer.Recite(const startBottles, takeDown: integer): string;
var
  index:integer;
  sb:TStringBuilder;
begin
  sb := TStringBuilder.Create;
  try
    for index := startBottles downto (startBottles - takeDown + 1) do begin
      if (index > 0) then begin
        sb.Append(Beer.NumBottles(index));
        sb.Append(' of beer on the wall, ');
        sb.Append(Beer.NumBottles(index));
        sb.Append(' of beer.\n');
        sb.Append('Take ');
        sb.Append(Beer.IsLastBottle(index));
        sb.Append(' down and pass it around, ');
        sb.Append(Beer.NumBottles(index-1));
        sb.Append(' of beer on the wall.\n');

        if (index <> (startBottles - takeDown + 1)) then
          sb.Append('\n');
      end
      else begin
        sb.Append('No more bottles of beer on the wall, no more bottles of beer.\n');
        sb.Append('Go to the store and buy some more, 99 bottles of beer on the wall.\n');
      end;
    end;
    Result := sb.ToString;
  finally
    FreeAndNil(sb);
  end;

end;

end.